controller/program/src/instructions/swap_exact_in.rs

```diff
@@ -0,0 +1,394 @@
+
+    let pool_trading_protocol_fee_bps = pool.trading_protocol_fee_bps;
+
+    let start_total_sol_value = pool.total_sol_value;
+
+    let inp_retval = cpi_lst_to_sol(
+        cpi,
+        abr,
+        *inp_calc_prog,
+        args.amount,
+        SvcIxAccountHandles::new(
+            NewSvcIxPreAccsBuilder::start()
+                .with_lst_mint(*ix_prefix.inp_lst_mint())
+                .build(),
+            inp_calc,
+        ),
+    )?;
+
+    let inp_sol_value = *inp_retval.start();
+    if inp_sol_value == 0 {
+        return Err(Inf1CtlCustomProgErr(Inf1CtlErr::ZeroValue).into());
+    }
+
+    let out_sol_val = cpi_price_exact_in(
+        cpi,
+        abr,
+        *pricing_prog,
+        PriceExactInIxArgs {
+            amt: args.amount,
+            sol_value: *inp_retval.start(),
+        },
+        PriceExactInIxAccountHandles::new(
+            NewPpIxPreAccsBuilder::start()
+                .with_input_mint(*ix_prefix.inp_lst_mint())
+                .with_output_mint(*ix_prefix.out_lst_mint())
+                .build(),
+            pricing,
+        ),
+    )?;
+
+    let out_retval = cpi_sol_to_lst(
+        cpi,
+        abr,
+        *out_calc_prog,
+        out_sol_val,
+        SvcIxAccountHandles::new(
+            NewSvcIxPreAccsBuilder::start()
+                .with_lst_mint(*ix_prefix.out_lst_mint())
+                .build(),
+            out_calc,
+        ),
+    )?;
+
+    if *out_retval.start() < args.limit {
+        return Err(Inf1CtlCustomProgErr(Inf1CtlErr::SlippageToleranceExceeded).into());
+    }
+
+    let inp_calc_retval = LstToSolRetVal(inp_retval);
+    let out_calc_retval = SolToLstRetVal(out_retval);
+    let pricing_retval = PricingRetVal(out_sol_val);
+
+    let quote = quote_exact_in(SwapQuoteArgs {
+        amt: args.amount,
+        out_reserves: out_lst_balance,
+        trading_protocol_fee_bps: pool_trading_protocol_fee_bps,
+        inp_calc: inp_calc_retval,
+        out_calc: out_calc_retval,
+        pricing: pricing_retval,
+        inp_mint: *abr.get(*ix_prefix.inp_lst_mint()).key(),
+        out_mint: *abr.get(*ix_prefix.out_lst_mint()).key(),
+    })
+    .map_err(|e| ProgramError::from(SwapQuoteProgErr(e)))?;
+
+    let inp_lst_token_program = *abr.get(inp_lst_token_program).key();
+    let inp_lst_decimals = RawMint::of_acc_data(abr.get(*ix_prefix.inp_lst_mint()).data())
+        .and_then(Mint::try_from_raw)
+        .map(|a| a.decimals())
+        .ok_or(INVALID_ACCOUNT_DATA)?;
+
+    let inp_lst_transfer_accs = NewTransferCheckedIxAccsBuilder::start()
+        .with_auth(*ix_prefix.signer())
+        .with_mint(*ix_prefix.inp_lst_mint())
+        .with_src(*ix_prefix.inp_lst_acc())
+        .with_dst(inp_pool_reserves)
+        .build();
+
+    cpi.invoke_fwd(
+        abr,
+        &inp_lst_token_program,
+        TransferCheckedIxData::new(args.amount, inp_lst_decimals).as_buf(),
+        inp_lst_transfer_accs.0,
+    )?;
```

Comment:
nit: you can use `invoke_fwd_handle` for programs that are in context
```suggestion
    let inp_lst_decimals = RawMint::of_acc_data(abr.get(*ix_prefix.inp_lst_mint()).data())
        .and_then(Mint::try_from_raw)
        .map(|a| a.decimals())
        .ok_or(INVALID_ACCOUNT_DATA)?;

    let inp_lst_transfer_accs = NewTransferCheckedIxAccsBuilder::start()
        .with_auth(*ix_prefix.signer())
        .with_mint(*ix_prefix.inp_lst_mint())
        .with_src(*ix_prefix.inp_lst_acc())
        .with_dst(inp_pool_reserves)
        .build();

    cpi.invoke_fwd_handle(
        abr,
        &inp_lst_token_program,
        TransferCheckedIxData::new(args.amount, inp_lst_decimals).as_buf(),
        inp_lst_transfer_accs.0,
    )?;
```